btrack/utils.py

```diff
@@ -221,6 +226,88 @@ def nans_idx(x):
     return data, properties, graph
 
 
+def napari_to_tracks(
+    data: npt.NDArray,
+    properties: Optional[dict[str, npt.ArrayLike]],
+    graph: Optional[dict[int, list[int]]],
+) -> list[btypes.Tracklet]:
+    """Convert napari Tracks to a list of Tracklets.
+
+    Parameters
+    ----------
+    data : array (N, D+1)
+        Coordinates for N points in D+1 dimensions. ID,T,(Z),Y,X. The first
+        axis is the integer ID of the track. D is either 3 or 4 for planar
+        or volumetric timeseries respectively.
+    properties : dict {str: array (N,)}
+        Properties for each point. Each property should be an array of length N,
+        where N is the number of points.
+    graph : dict {int: list}
+        Graph representing associations between tracks. Dictionary defines the
+        mapping between a track ID and the parents of the track. This can be
+        one (the track has one parent, and the parent has >=1 child) in the
+        case of track splitting, or more than one (the track has multiple
+        parents, but only one child) in the case of track merging.
+
+    Returns
+    -------
+    tracks : list[btypes.Tracklet]
+        A list of tracklet objects from BayesianTracker.
+
+    """
+
+    if data.shape[1] == Dimensionality.FIVE:
+        track_id, t, z, y, x = data.T
+    elif data.shape[1] == Dimensionality.FOUR:
+        track_id, t, y, x = data.T
+        z = np.zeros_like(x)
+    else:
+        raise ValueError(
+            "Data must have either 4 (ID, t, y, x) or 5 (ID, t, z, y, x) columns, "
+            f"not {data.shape[1]}"
+        )
+
+    # Create all PyTrackObjects
+    objects_dict = {
+        "ID": properties.get("refs", np.arange(track_id.size)),
+        "t": t,
+        "x": x,
+        "y": y,
+        "z": z,
+        "dummy": properties.get("dummy", np.full_like(track_id, fill_value=False)),
+        "label": properties.get(
+            "state", np.full_like(track_id, fill_value=constants.States.NULL)
+        ),
+    }
+    track_objects = objects_from_dict(objects_dict)
+
+    # Create all Tracklets
+    tracklets = []
+    for track in np.unique(track_id).astype(int):
+        # Create tracklet
+        track_indices = np.argwhere(track_id == track).ravel()
+        track_data = [track_objects[i] for i in track_indices]
+        parent = graph.get(track, [track])[0]
+        children = [child for (child, parents) in graph.items() if track in parents]
+        tracklet = Tracklet(
+            ID=track,
+            data=track_data,
+            parent=parent,
+            children=children,
+        )
+
+        # Determine root tracklet
+        tracklet.root = parent
+        tracklet.generation = 0 if tracklet.root == track else 1
+        while tracklet.root in graph:
+            tracklet.root = graph[tracklet.root][0]
+            tracklet.generation += 1
```

Comment:
I'm going to assume the tree traversal logic is good here.

Reply:
I _think_ so - at least with the test dataset it's correct